synapse/metrics/background_process_metrics.py

```diff
@@ -197,7 +199,15 @@ def run():
                 _background_processes.setdefault(desc, set()).add(proc)
 
             try:
-                yield func(*args, **kwargs)
+                # We ensureDeferred here to handle coroutines
+                result = func(*args, **kwargs)
+
+                # We need this check because ensureDeferred doesn't like when
+                # func doesn't return a Deferred or coroutine.
```

Comment:
riiight... but why are we calling `run_as_background_process` on a thing which doesn't return a coroutine or a deferred?

Reply:
I get a `RuntimeWarning: coroutine 'ClientIpStore._prune_old_user_ips' was never awaited` if I don't do this

Reply:
https://github.com/matrix-org/synapse/blob/master/synapse/util/caches/expiringcache.py#L75-L80 is an example where we are periodically running code that isn't async. I *think* its still legit for it to have a logcontext set in that case?

Reply:
oh, right. Yes, I guess it is.